### PR TITLE
LIS3DH accelerometer driver

### PR DESCRIPTION
**Pull request by gebart**
I wrote a driver for the ST Micro LIS3DH accelerometer. It only provides basic functionality of the sensor. 
Polling in streaming or FIFO modes. No interrupts (INT1, INT2).

The sensor name is LIS3DH MEMS digital output motion sensor ultra low-power high performance 3-axes "nano" accelerometer

Manufacturer website: http://www.st.com/web/catalog/sense_power/FM89/SC444/PF250725
Data sheet: http://www.st.com/st-web-ui/static/active/en/resource/technical/document/datasheet/CD00274221.pdf

I have executed the test application on my Mulle board and it works there.

I based the structure of the driver and the test application on the design of the L3G4200 gyroscope driver.

### CODE REVIEW

File: drivers/include/lis3dh.h [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+  LIS3DH_ODR_50Hz      = (0x04 << LIS3DH_CTRL_REG1_ODR_SHIFT),
+  LIS3DH_ODR_100Hz     = (0x05 << LIS3DH_CTRL_REG1_ODR_SHIFT),
+  LIS3DH_ODR_200Hz     = (0x06 << LIS3DH_CTRL_REG1_ODR_SHIFT),
+  LIS3DH_ODR_400Hz     = (0x07 << LIS3DH_CTRL_REG1_ODR_SHIFT),
+  LIS3DH_ODR_LP1600Hz  = (0x08 << LIS3DH_CTRL_REG1_ODR_SHIFT),
+  LIS3DH_ODR_NP1250Hz_LP5000HZ = (0x09 << LIS3DH_CTRL_REG1_ODR_SHIFT)
+} lis3dh_odr_t;
+
+/**
+ * Low power mode enable. Default value: 0
```
Comment (OlegHahm) [inline drivers/include/lis3dh.h]: Maybe use `@name` for this and the following group of macros.

File: drivers/include/lis3dh.h [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+ *
+ * @return                  0 on success
+ * @return                  -1 on error
+ */
+int lis3dh_set_aux_adc(const lis3dh_t *dev, const uint8_t enable, const uint8_t temperature);
+
+/**
+ * @brief Enable/disable accelerometer axes.
+ *
+ * @param[in]  axes         An OR-ed combination of LIS3DH_AXES_X,
```
Comment (OlegHahm) [inline drivers/include/lis3dh.h]: `@param[in] dev` is missing here and for the following functions.

File: drivers/lis3dh/lis3dh.c [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+  return 0;
+}
+
+int lis3dh_read_xyz(const lis3dh_t *dev, lis3dh_data_t *acc_data)
+{
+  /* Set READ MULTIPLE mode */
+  static const uint8_t addr = (LIS3DH_REG_OUT_X_L | LIS3DH_SPI_READ_MASK | LIS3DH_SPI_MULTI_MASK);
+
+  /* Acquire exclusive access to the bus. */
+  //~ spi_acquire_bus(dev->spi);
```
Comment (PeterKietzmann) [inline drivers/lis3dh/lis3dh.c]: his will be `spi_aquire(dev->spi);` after #2290 is merged

File: drivers/lis3dh/lis3dh.c [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+  /* Perform the transaction */
+  gpio_clear(dev->cs);
+  if (spi_transfer_regs(dev->spi, addr, NULL, (char*)acc_data, sizeof(lis3dh_data_t)) != sizeof(lis3dh_data_t)) {
+    /* Transfer error */
+    return -1;
+  }
+  gpio_set(dev->cs);
+  /* Release the bus for other threads. */
+  //~ spi_release_bus(dev->spi);
```
Comment (PeterKietzmann) [inline drivers/lis3dh/lis3dh.c]: This will be spi_release(dev->spi); after #2290 is merged

File: drivers/lis3dh/lis3dh.c [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+ *
+ * @return                  0 on success
+ * @return                  -1 on error
+ */
+static int lis3dh_read_regs(const lis3dh_t *dev, const lis3dh_reg_t reg, const uint8_t len, uint8_t* buf) {
+  /* Set READ MULTIPLE mode */
+  uint8_t addr = (reg & LIS3DH_SPI_ADDRESS_MASK) | LIS3DH_SPI_READ_MASK | LIS3DH_SPI_MULTI_MASK;
+
+  /* Acquire exclusive access to the bus. */
+  //~ spi_acquire_bus(dev->spi);
```
Comment (PeterKietzmann) [inline drivers/lis3dh/lis3dh.c]: `spi_acuire();`

File: drivers/lis3dh/lis3dh.c [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+  /* Perform the transaction */
+  gpio_clear(dev->cs);
+  if (spi_transfer_regs(dev->spi, addr, NULL, (char*)buf, len) < 0) {
+    /* Transfer error */
+    return -1;
+  }
+  gpio_set(dev->cs);
+  /* Release the bus for other threads. */
+  //~ spi_release_bus(dev->spi);
```
Comment (PeterKietzmann) [inline drivers/lis3dh/lis3dh.c]: `spi_release();`

File: drivers/lis3dh/lis3dh.c [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+ *
+ * @return                  0 on success
+ * @return                  -1 on error
+ */
+static int lis3dh_write_reg(const lis3dh_t *dev, const lis3dh_reg_t reg, const uint8_t value) {
+  /* Set WRITE SINGLE mode */
+  uint8_t addr = (reg & LIS3DH_SPI_ADDRESS_MASK) | LIS3DH_SPI_WRITE_MASK | LIS3DH_SPI_SINGLE_MASK;
+
+  /* Acquire exclusive access to the bus. */
+  //~ spi_acquire_bus(dev->spi);
```
Comment (PeterKietzmann) [inline drivers/lis3dh/lis3dh.c]: like above

File: drivers/lis3dh/lis3dh.c [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+  /* Perform the transaction */
+  gpio_clear(dev->cs);
+  if (spi_transfer_reg(dev->spi, addr, value, NULL) < 0) {
+    /* Transfer error */
+    return -1;
+  }
+  gpio_set(dev->cs);
+  /* Release the bus for other threads. */
+  //~ spi_release_bus(dev->spi);
```
Comment (PeterKietzmann) [inline drivers/lis3dh/lis3dh.c]: like above

**Comment by gebart** [comment]
Updated Doxygen according to comments, replaced disabled `spi_acquire_bus`->`spi_acquire`, replaced disabled `spi_release`->`spi_release`

**Comment by PeterKietzmann** [comment]
Nice! Do you know if this is compatible with the LIS302DL or LIS3DSH which are mounted on the stm32f4discovery board?

**Comment by gebart** [comment]
They are similar, but the LIS3DSH has a different response to the WHO_AM_I query.

The basic accelerometer functionality seem the same between the LIS3DH and LIS3DSH, but the LIS3DSH has some different registers for configuration and does not have the extra 3 auxiliary ADC channels on the LIS3DH. The changes seem too large to be easily made compatible, this driver can be used as a base for a future LIS3DSH driver.
The LIS302DL also seem similar, but not the same either.

**Comment by haukepetersen** [comment]
As far as I see it you return raw data when reading the sensor values, right? I would actually vote for returning a physical value (as e.g. mili G). This is what we have done for most the other sensors.

When I use a sensor, even more importantly when I do sensor fusion, that's the values I am interested in and as a user I don't want to read the data sheet to figure out how to translate the raw values...

**Comment by gebart** [comment]
@haukepetersen okay, I will add this scaling.

**Comment by PeterKietzmann** [comment]
```
ERROR: The following new files generate doxygen warnings:

drivers/include/lis3dh.h
```

**Comment by gebart** [comment]
@haukepetersen After reading the data sheet again I noticed that it states that the scale is 1 mg per digit at +/- 2G, 2 mg per digit at +/- 4G etc. The values are left-aligned, meaning that we do not need to scale depending on which sensor full scale has been selected. In ST Application Note AN3308 *LIS3DH: MEMS digital output motion sensor
ultra low-power high performance 3-axis “nano” accelerometer*, **3.2.3
Example of acceleration data, Table 6** however, it says that 0x400 equals 1 G and 0xC000 equals -1 G, which would imply that 1 G equals 1024 LSB, not 1000 as the data sheet says.

I do not have the device with me at the moment, I need to do some experiments when I get back in order to verify the correct values, if it is at all noticeable, the difference is pretty small and close to the noise level.

**Comment by OlegHahm** [comment]
Travis fail is unrelated.

**Comment by gebart** [comment]
Added scaling to milli G.

From my experiments it seems like the scale at 2G setting is such that 1 G is approx 16384, not 16000, so the app note seem correct.

**Comment by PeterKietzmann** [comment]
@gebart I have no hardware to test. Anyone has? Everything looks valid to me, in results I think I can trust you. How about squashing?

**Comment by gebart** [comment]
@PeterKietzmann I have refactored to use `int16_t scale` in the driver struct for caching the scale factor.

Squash and go?

**Comment by PeterKietzmann** [comment]
Yes looks good. Please squash, then let's see what Travis says.

**Comment by PeterKietzmann** [comment]
As nobody has this sensor available I'll trust you that it works:-). Everything looks fine and the unclear points had been discussed. ACK

**Pull request merged**